### ISSUE TITLE
Smarter settings for default init parameters in SP/TP/encoders

### ISSUE THREAD

**Issue by breznak** [numenta/nupic-legacy#1343]
- [ ] better defaults and 
- [ ] improve "checks" and relations between the paramaters
- [ ] ideally have versions with "empty" () possible constructor

Suggested guide-lines: 
- use relative rather than absolute values
- find linked parameters and replace with only 1 param and compute the others

TODO: 
- [ ] review Spatial Pooler
- [ ] review Temporal Pooler
- [ ] review Encoders
  - [ ] `w` param should be relative to `n`, not abs. value
  - [ ] add `forced` flag that disables any automatic settings behind the scenes.

QUESTIONS:
- [ ] should we prefer proportional (%) values?

**Comment by breznak** [comment]
see https://github.com/numenta/nupic.research/issues/207#issuecomment-117985501

**Comment by cogmission** [comment]
Hi,

Can I direct you to the discussion about a tool I proposed? @breznak
@subutai @rhyolight

http://lists.numenta.org/pipermail/nupic-hackers_lists.numenta.org/2014-September/002696.html

On Thu, Jul 2, 2015 at 4:53 AM, breznak <notifications@github.com> wrote:

> see numenta/nupic.research#207 (comment)
> <https://github.com/numenta/nupic.research/issues/207#issuecomment-117985501>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numenta/nupic/issues/1343#issuecomment-117985758>.
>



-- 
*With kind regards,*

David Ray
Java Solutions Architect

*Cortical.io <http://cortical.io/>*
Sponsor of:  HTM.java <https://github.com/numenta/htm.java>

d.ray@cortical.io
http://cortical.io

**Comment by breznak** [comment]
@cogmission hi David, thanks for the link! I agree with the idea but have some suggestions: 
- :+1: for gui/web app with sliders
  - I think just the 2 params: input size & columns size are not enough - I'd like (all) params and their relationships
  - maybe related: https://github.com/numenta/nupic/issues/2251 
- this issue is about reviewing existing defaults (some are quite bad, like TP columns=500 !)
- when you assume to modify just the 2 params - why not updating the best settings internally in SP/TP already? that is, less params are user given and more relations between params should be encoded

**Comment by cogmission** [comment]
@breznak

I think you're misunderstanding me. I proposed a "sheet" with all configs
for a given algorithm - as a user changes one, the others change
automatically to keep within certain sanity policies. They would all be
"bound" to each other and react accordingly. Matt suggested making it in
NodeJS and Subutai said he would help with developing the interrelated
relationships and policies for interrelatedness.

I'm not sure where you got 2 configs from? Look at the whole thread...
(it's not that long)

On Thu, Jul 2, 2015 at 6:27 AM, breznak <notifications@github.com> wrote:

> @cogmission <https://github.com/cogmission> hi David, thanks for the
> link! I agree with the idea but have some suggestions:
>
>    - [image: :+1:] for gui/web app with sliders
>       - I think just the 2 params: input size & columns size are not
>       enough - I'd like (all) params and their relationships
>       - maybe related: #2251
>       <https://github.com/numenta/nupic/issues/2251>
>     - this issue is about reviewing existing defaults (some are quite
>    bad, like TP columns=500 !)
>    - when you assume to modify just the 2 params - why not updating the
>    best settings internally in SP/TP already? that is, less params are user
>    given and more relations between params should be encoded
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numenta/nupic/issues/1343#issuecomment-118004525>.
>



-- 
*With kind regards,*

David Ray
Java Solutions Architect

*Cortical.io <http://cortical.io/>*
Sponsor of:  HTM.java <https://github.com/numenta/htm.java>

d.ray@cortical.io
http://cortical.io

**Comment by cogmission** [comment]
@breznak

Oh I see where you got that (the reference to two variables). That was just
to arrive at the initial state. So you would begin with an initial
condition, then all the other vars would "adapt" to suit. Then the user
could adjust any of the others (other than the original two) - and all
others would change to accommodate the new setting for the current variable
the user was adjusting?

On Thu, Jul 2, 2015 at 6:44 AM, cogmission (David Ray) <
cognitionmission@gmail.com> wrote:

> @breznak
>
> I think you're misunderstanding me. I proposed a "sheet" with all configs
> for a given algorithm - as a user changes one, the others change
> automatically to keep within certain sanity policies. They would all be
> "bound" to each other and react accordingly. Matt suggested making it in
> NodeJS and Subutai said he would help with developing the interrelated
> relationships and policies for interrelatedness.
>
> I'm not sure where you got 2 configs from? Look at the whole thread...
> (it's not that long)
>
> On Thu, Jul 2, 2015 at 6:27 AM, breznak <notifications@github.com> wrote:
>
>> @cogmission <https://github.com/cogmission> hi David, thanks for the
>> link! I agree with the idea but have some suggestions:
>>
>>    - [image: :+1:] for gui/web app with sliders
>>       - I think just the 2 params: input size & columns size are not
>>       enough - I'd like (all) params and their relationships
>>       - maybe related: #2251
>>       <https://github.com/numenta/nupic/issues/2251>
>>     - this issue is about reviewing existing defaults (some are quite
>>    bad, like TP columns=500 !)
>>    - when you assume to modify just the 2 params - why not updating the
>>    best settings internally in SP/TP already? that is, less params are user
>>    given and more relations between params should be encoded
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/numenta/nupic/issues/1343#issuecomment-118004525>.
>>
>
>
>
> --
> *With kind regards,*
>
> David Ray
> Java Solutions Architect
>
> *Cortical.io <http://cortical.io/>*
> Sponsor of:  HTM.java <https://github.com/numenta/htm.java>
>
> d.ray@cortical.io
> http://cortical.io
>



-- 
*With kind regards,*

David Ray
Java Solutions Architect

*Cortical.io <http://cortical.io/>*
Sponsor of:  HTM.java <https://github.com/numenta/htm.java>

d.ray@cortical.io
http://cortical.io

**Comment by breznak** [comment]
> I proposed a "sheet" with all configs for a given algorithm - as a user changes one, the others change automatically to keep within certain sanity policies. They would all be "bound" to each other and react accordingly.

Why don't do it in the algorithm already then? Apply the "bounding" to constructor arguments and hook to any `setXXX(val)` calls? That way swarming & all user defined parameters work in optimal settings.

**Comment by cogmission** [comment]
I can think of two things.

1. Having a "tool" like this could allow a user to "see" how the parameters
relate; and eventually learn about them and get an intuitive sense for
their meaning.
2. We need to be careful about automatically adjusting things (inserting
side effects) without the programmer being aware. There are always outliers
where someone might want some "weird" configuration not expected or
anticipated beforehand?



On Thu, Jul 2, 2015 at 7:13 AM, breznak <notifications@github.com> wrote:

> I proposed a "sheet" with all configs for a given algorithm - as a user
> changes one, the others change automatically to keep within certain sanity
> policies. They would all be "bound" to each other and react accordingly.
>
> Why don't do it in the algorithm already then? Apply the "bounding" to
> constructor arguments and hook to any setXXX(val) calls? That way
> swarming & all user defined parameters work in optimal settings.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numenta/nupic/issues/1343#issuecomment-118011544>.
>



-- 
*With kind regards,*

David Ray
Java Solutions Architect

*Cortical.io <http://cortical.io/>*
Sponsor of:  HTM.java <https://github.com/numenta/htm.java>

d.ray@cortical.io
http://cortical.io

**Comment by breznak** [comment]
> We need to be careful about automatically adjusting things (inserting side effects) without the programmer being aware. There are always outliers where someone might want some "weird" configuration not expected or anticipated beforehand?

good point. added `forced` flag to description, function same as in encoders.

**Comment by subutai** [comment]
I like the idea of better default values. I just did a quick check and saw that they are off quite a bit for the SP and TP. For example, having only 10 winners for the SP is just plain wrong. I would recommend using the hotgym values as defaults as they have been well tested in a number of scenarios.  

A tool could be nice too. It can use the SDR math to provide good suggestions for many situations. Perhaps this should start in the community research repo, and then if it looks useful in general we can move it into NuPIC?

**Comment by breznak** [comment]
@subutai agreed. I suggest splitting in 2: 
- [ ] immediate fix for way-off default values (as from our example in https://github.com/numenta/nupic.research/issues/207#issuecomment-117985501 )
- [ ] `community.research` to explore more relations between params, limit number of user-set params & create (GUI) tool to fiddle with the settings.

**Comment by subutai** [comment]
@breznak The plan sounds good to me.

**Comment by cogmission** [comment]
@subutai @breznak Me too! :-)
